### PR TITLE
Reduce Cluster memory utilization by avoiding Metadata instances when not needed

### PR DESCRIPTION
**Pull request by yurykats**
Reduce Cluster memory utilization by avoiding Metadata instances when not needed

Currently an instance of raw Metadata and an instance of TypedMetadata are created in each cluster, even if no metadata is configured. This PR switches instances to unique_ptr and only creates Metadata objects when actually configured. Otherwise a singleton default metadata is used.

This is a step towards more memory efficient config data structures. Issue #24154.

A similar change was made to the Route in #24309. This PR is consistent with the approach applied there.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by adisuissa — COMMENTED** [review]
Thanks!

I'm not sure these can be easily separated, as the typed and non-typed are intertwined and one can access data of another.
I think that there's something that can be done to reduce memory, and that's splitting the typed and non-typed so at least Envoy doesn't keep the same data in both. @DiazAlan and I discussed this a while ago, and I'm not sure if a PR was submitted.

**Comment by yurykats** [comment]
> I'm not sure these can be easily separated, as the typed and non-typed are intertwined and one can access data of another. 

What do you mean by "access data of another"?  They are created from the same config proto, but once created, they each have their own copy of metadata, in different format. If there is metadata in the config to begin with, there is no point in having those class members, is there?

> I think that there's something that can be done to reduce memory ... so at least Envoy doesn't keep the same data in both. @DiazAlan and I discussed this a while ago, and I'm not sure if a PR was submitted.

Yes, it would be a good improvement, but it would also be a breaking API change, since users can currently access the same metadata both ways. We can follow up with additional refactoring here, if we find a way.

**Comment by adisuissa** [comment]
> What do you mean by "access data of another"? They are created from the same config proto, but once created, they each have their own copy of metadata, in different format. If there is metadata in the config to begin with, there is no point in having those class members, is there?
> 
The current implementation allows accessing typed-metadata similar to regular metadata which could lead to undesired access patterns ([example](https://github.com/envoyproxy/envoy/blob/02204331c80296840082c4b13a82af5e88e3934c/source/extensions/filters/http/gcp_authn/gcp_authn_filter.cc#L49-L50).

That said, I see the benefit of this PR, and agree that it is decoupled from my discussion with @DiazAlan.

Let me review this PR.

File: source/common/upstream/upstream_impl.h [hunk 1]
```diff
@@ -869,8 +869,22 @@ class ClusterInfoImpl : public ClusterInfo,
...
+    return DefaultMetadata::get();
+  }
+  using ClusterTypedMetadata = Envoy::Config::TypedMetadataImpl<ClusterTypedMetadataFactory>;
+  const Envoy::Config::TypedMetadata& typedMetadata() const override {
+    if (typed_metadata_ != nullptr) {
+      return *typed_metadata_;
+    }
+    static const ClusterTypedMetadata* defaultTypedMetadata =
+        new ClusterTypedMetadata(DefaultMetadata::get());
+    return *defaultTypedMetadata;
```
Comment (adisuissa) [inline source/common/upstream/upstream_impl.h:886]: nit: consider using [CONSTRUCT_ON_FIRST_USE](https://github.com/envoyproxy/envoy/blob/02204331c80296840082c4b13a82af5e88e3934c/source/common/common/macros.h#L32).

**Review by adisuissa — APPROVED** [review]
LGTM, left a small nit for consideration.
Thanks!

File: source/common/upstream/upstream_impl.h (hunk 1 above)
Comment (yurykats) [reply source/common/upstream/upstream_impl.h]: Done

**Comment by yurykats** [comment]
> The current implementation allows accessing typed-metadata similar to regular metadata which could lead to undesired access patterns ([example]

Yes, indeed. But changing this would require a breaking API change, no?

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

**Review by lizan — APPROVED** [review]

**Pull request merged**